src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -733,162 +765,215 @@ DatabaseShardImp::onChildrenStopped()
 void
 DatabaseShardImp::import(Database& source)
 {
+    assert(init_);
+
+    // Only the application local node store can be imported
+    if (&source != &app_.getNodeStore())
     {
-        std::lock_guard lock(mutex_);
-        assert(init_);
+        assert(false);
+        JLOG(j_.error()) << "invalid source database";
+        return;
+    }
+
+    std::uint32_t earliestIndex;
+    std::uint32_t latestIndex;
+    {
+        auto loadLedger = [&](bool ascendSort =
+                                  true) -> boost::optional<std::uint32_t> {
+            std::shared_ptr<Ledger> ledger;
+            std::uint32_t ledgerSeq;
+            std::tie(ledger, ledgerSeq, std::ignore) = loadLedgerHelper(
+                "WHERE LedgerSeq >= " + std::to_string(earliestLedgerSeq()) +
+                    " order by LedgerSeq " + (ascendSort ? "asc" : "desc") +
+                    " limit 1",
+                app_,
+                false);
+            if (!ledger || ledgerSeq == 0)
+            {
+                JLOG(j_.error()) << "no suitable ledgers were found in"
+                                    " the SQLite database to import";
+                return boost::none;
+            }
+            return ledgerSeq;
+        };
+
+        // Find earliest ledger sequence stored
+        auto ledgerSeq{loadLedger()};
+        if (!ledgerSeq)
+            return;
+        earliestIndex = seqToShardIndex(*ledgerSeq);
 
-        // Only the application local node store can be imported
-        if (&source != &app_.getNodeStore())
+        // Consider only complete shards
+        if (ledgerSeq != firstLedgerSeq(earliestIndex))
+            ++earliestIndex;
+
+        // Find last ledger sequence stored
+        ledgerSeq = loadLedger(false);
+        if (!ledgerSeq)
+            return;
+        latestIndex = seqToShardIndex(*ledgerSeq);
+
+        // Consider only complete shards
+        if (ledgerSeq != lastLedgerSeq(latestIndex))
+            --latestIndex;
+
+        if (latestIndex < earliestIndex)
         {
-            assert(false);
-            JLOG(j_.error()) << "invalid source database";
+            JLOG(j_.error()) << "no suitable ledgers were found in"
+                                " the SQLite database to import";
             return;
         }
+    }
+
+    auto numHistShards = numHistoricalShards(std::lock_guard(mutex_));
+    auto const sweepInterval = std::chrono::milliseconds(500);
+
+    // Sweep the source and family caches to conserve
+    // memory
+    ThreadWrapper mainSweeper(
+        [&source, sweepInterval, &family = app_.getNodeFamily()] {
+            source.sweep();
+            family.sweep();
+            std::this_thread::sleep_for(sweepInterval);
+        });
```

Comment:
>     1. What happens if the sweep is still in process when the call to `import()` completes and the destructor of `mainSweeper` runs.  Is that okay?  It might be good to add a comment so casual readers know you've thought about this problem.
> 

In this event the `mainSweeper` destructor blocks until the current iteration of the `ThreadWrapper::run` loop completes. I added a comment to the `ThreadWrapper` destructor.

>     2. It's becoming apparent that `import()` can be a very long running method.  Since `DatabaseShardImp` is a `Stoppable`, I'd expect this method to be sprinkled with calls to `isStopping()` to allow an early exit.  I'm not seeing any.  What are the plans for handling early termination to allow a graceful shutdown while `import()` is running?

Since the `Stoppable` interface isn't running at this point in the code, this approach to graceful stops isn't feasible. I've just been killing the process manually.

Reply:
It's worth mentioning in a comment that `Stoppable` isn't running so you can't check it.

I learned today that @thejohnfreeman is intending to yank all of `Stoppable`?  I've not seen anything about what they have in mind.  But you might see if you can have a chat and learn what the intended replacement is and how you might integrate with it?

Reply:
I just posted my planned changes in #3741. You can see how `DatabaseShardImp` is impacted there. I tried to prepare a short answer by skimming this PR, but I think it'll take a deep review for me to understand what is going on here.

Reply:
Done (added comment)